OP-TEE Benchmark

igoropaniuk opened the pull request:
Related PRs:

Benchmark app:
https://github.com/linaro-swg/optee_benchmark/pull/1

Changes in OP-TEE
https://github.com/linaro-swg/linux/pull/38
https://github.com/OP-TEE/build/pull/124
https://github.com/OP-TEE/optee_os/pull/1365
https://github.com/OP-TEE/optee_client/pull/79

Commit:
OP-TEE Benchmark **not for mainline**

Add Benchmark support

Reviewed-by: Joakim Bech <joakim.bech@linaro.org>
Signed-off-by: Igor Opaniuk <igor.opaniuk@linaro.org>

jenswikl commented on drivers/tee/optee/call.c:
```diff
@@ -21,13 +21,19 @@
 	bool completed;
 };
 
+
```
Empty line not needed

jenswikl commented on drivers/tee/optee/call.c:
```diff
@@ -332,6 +342,15 @@ int optee_invoke_func(struct tee_context *ctx, struct tee_ioctl_invoke_arg *arg,
 
+#ifdef CONFIG_OPTEE_BENCHMARK
+	if (arg->func == BENCHMARK_CMD_REGISTER_MEMREF) {
+		bench_ts_buf = (struct tee_time_buf *) tee_shm_get_va(
```
Cast not needed

jenswikl commented on drivers/tee/optee/call.c:
```diff
@@ -332,6 +342,15 @@ int optee_invoke_func(struct tee_context *ctx, struct tee_ioctl_invoke_arg *arg,
+		bench_ts_buf = (struct tee_time_buf *) tee_shm_get_va(
+						param[0].u.memref.shm, 0);
+	} else if (arg->func == BENCHMARK_CMD_UNREGISTER) {
+		bench_ts_buf = 0;
```
Please use `NULL` instead of `0` when assigning pointers.

jenswikl commented on drivers/tee/optee/core.c:
```diff
@@ -567,10 +568,55 @@ static struct platform_driver optee_driver = {
 	.remove = optee_remove,
 };
 
+static void enable_cpu_perf_counters(void *data)
```
This would be better to have in a separate file to minimize the risk of future conflicts.

jenswikl commented on drivers/tee/optee/optee_bench.h:
```diff
@@ -0,0 +1,134 @@
+#define OPTEE_BENCH_KMOD	0x20000000
+
+/* storing timestamps */
+struct tee_time_st {
```
Proper prefix is `optee_`

jenswikl commented on drivers/tee/optee/optee_bench.h:
```diff
@@ -0,0 +1,134 @@
+#define OPTEE_BENCH_MAX_STAMPS	10
+#define OPTEE_BENCH_RB_SIZE (sizeof(struct tee_time_buf) + \
+		sizeof(struct tee_time_st) * OPTEE_BENCH_MAX_STAMPS)
+#define OPTEE_BENCH_DEF_PARAM		4
```
Unused?

jenswikl commented on drivers/tee/optee/optee_bench.h:
```diff
@@ -0,0 +1,134 @@
+
+/* max amount of timestamps */
+#define OPTEE_BENCH_MAX_STAMPS	10
+#define OPTEE_BENCH_RB_SIZE (sizeof(struct tee_time_buf) + \
```
Unused?

jenswikl commented on drivers/tee/optee/optee_bench.h:
```diff
@@ -0,0 +1,134 @@
+
+/* aarch64 */
+#define OPTEE_BENCH_ARMV8_PMCR_MASK	0x3f
+#define OPTEE_BENCH_ARMV8_PMCR_E	(1 << 0) /* Enable all counters */
```
Use `BIT()` macro.

jenswikl commented on drivers/tee/optee/optee_bench.h:
```diff
@@ -0,0 +1,134 @@
+{
+	u64 ts_i;
+
+	if ((cmd_id & BENCHMARK_CMD(0)) == BENCHMARK_CMD(0))
```
What's the purpose of this?

igoropaniuk replied:
To avoid adding timestamps for calls to Benchmark PTA
similar comment https://github.com/OP-TEE/optee_client/pull/79/files#r103206052

jenswikl replied:
And it's not harmless to log timestamps for calls to Benchmark PTA?

jenswikl commented on drivers/tee/optee/optee_bench.h:
```diff
@@ -0,0 +1,134 @@
+#endif
+
+#ifdef CONFIG_OPTEE_BENCHMARK
+extern struct tee_time_buf *bench_ts_buf;
```
Please use a proper prefix

jenswikl commented on drivers/tee/optee/rpc.c:
```diff
@@ -313,15 +314,36 @@ static void handle_rpc_func_cmd_shm_free(struct tee_context *ctx,
+	switch(type) {
+		case OPTEE_MSG_RPC_CMD_BENCH_REG_NEW:
+			size = arg->params[0].u.value.c;
+			optee_bench_ts_global = (struct optee_ts_global *)
```
cast is not needed.

jenswikl commented on drivers/tee/optee/rpc.c:
```diff
@@ -313,15 +314,36 @@ static void handle_rpc_func_cmd_shm_free(struct tee_context *ctx,
+		case OPTEE_MSG_RPC_CMD_BENCH_REG_NEW:
+			size = arg->params[0].u.value.c;
+			optee_bench_ts_global = (struct optee_ts_global *)
+				ioremap(arg->params[0].u.value.b, size);
```
Use `memremap()` instead.

jenswikl commented on drivers/tee/optee/rpc.c:
```diff
@@ -313,15 +314,36 @@ static void handle_rpc_func_cmd_shm_free(struct tee_context *ctx,
+			break;
+		case OPTEE_MSG_RPC_CMD_BENCH_REG_DEL:
+			if (optee_bench_ts_global)
+				iounmap(optee_bench_ts_global);
```
`memunmap()`

jenswikl commented on drivers/tee/optee/rpc.c:
```diff
@@ -313,15 +314,36 @@ static void handle_rpc_func_cmd_shm_free(struct tee_context *ctx,
+			if (optee_bench_ts_global)
+				iounmap(optee_bench_ts_global);
+
+			optee_bench_ts_global = 0;
```
`s/0/NULL/`

igoropaniuk commented on drivers/tee/optee/bench.c:
```diff
@@ -0,0 +1,153 @@
+
+struct optee_ts_global *optee_bench_ts_global;
+
+static inline u32 armv8pmu_pmcr_read(void)
```
should be moved also to architecture-dependent dirs

igoropaniuk commented on drivers/tee/optee/optee_bench.h:
```diff
@@ -0,0 +1,58 @@
+
+#define OPTEE_BENCH_DIVIDER 		64
+/* max amount of timestamps */
+#define OPTEE_BENCH_MAX_STAMPS	32
```
I'll move this definition (`OPTEE_BENCH_MAX_STAMPS`) to common.mk in build, it's should be easy to change it in all sublayers

igoropaniuk commented:
Fix: 
Facing an issue `[  108.180621] Unhandled fault: implementation fault (unsupported exclusive) (0x92000035) at 0x0000007f7c474938` (from time to time) on RPi3

jockebech commented on drivers/tee/optee/bench.c:
```diff
@@ -0,0 +1,153 @@
+		u64 val = 0;
+
+		asm volatile("mrs %0, pmcr_el0" : "=r"(val));
+		return (u32)val;
```
So how does this work? Casting the 64-bit value to a 32-bit value here.

jockebech commented on drivers/tee/optee/bench.c:
```diff
@@ -0,0 +1,153 @@
+	cur_cpu = get_cpu();
+	/* fill timestamp data */
+	if (cur_cpu >= optee_bench_ts_global->cores)
+		return;
```
Don't we want to report some kind of error here. The function is void and nothing is being said here. I.e, we can silently jump out of this function when things goes wrong.

etienne-lms replied:
major: Call `put_cpu()` before returning.

jockebech commented on drivers/tee/optee/rpc.c:
```diff
@@ -310,6 +312,45 @@ static void handle_rpc_func_cmd_shm_free(struct tee_context *ctx,
 	arg->ret = TEEC_SUCCESS;
 }
 
+static void handle_rpc_func_cmd_bench_reg_buffer(struct optee_msg_arg *arg)
```
That's a long function name .... `handle_rpc_bm_reg` ? `handle_rpc_bm_regbuf`?

jockebech commented on drivers/tee/optee/rpc.c:
```diff
@@ -310,6 +312,45 @@ static void handle_rpc_func_cmd_shm_free(struct tee_context *ctx,
+			memunmap(optee_bench_ts_global);
+
+		optee_bench_ts_global = NULL;
+			break;
```
If it's not the GitHub gui that fools me, then this break should move back one tab.

etienne-lms commented on drivers/tee/optee/bench.c:
```diff
@@ -0,0 +1,153 @@
+	/* Disable counter overflow interrupts */
+	asm volatile("mcr p15, 0, %0, c9, c12, 1" :: "r"(OPTEE_BENCH_DEF_OVER));
+#endif
+
```
minor: empty line

etienne-lms commented on drivers/tee/optee/bench.c:
```diff
@@ -0,0 +1,153 @@
+	/* Disable EL0 access to PMU counters. */
+	asm volatile("mcr p15, 0, %0, c9, c14, 0" :: "r"(0));
+#endif
+
```
minor: empty line

etienne-lms commented on drivers/tee/optee/bench.c:
```diff
@@ -0,0 +1,153 @@
+	cpu_buf = &optee_bench_ts_global->cpu_buf[cur_cpu];
+	ts_i = __sync_fetch_and_add(&cpu_buf->head, 1);
+	ts_data.cnt = read_ccounter();
+	ts_data.addr = (uintptr_t) ret_addr;
```
minor: remove no space before cast'ed label.

etienne-lms commented on drivers/tee/optee/bench.c:
```diff
@@ -0,0 +1,153 @@
+/* stabs */
+void optee_bm_enable(void) {}
+void optee_bm_disable(void) {}
+void optee_bm_timestamp(void) {}
```
minor: could be inlined in **optee_bench.h**

etienne-lms commented on drivers/tee/optee/rpc.c:
```diff
@@ -307,6 +309,46 @@ static void handle_rpc_func_cmd_shm_free(struct tee_context *ctx,
+			memunmap(optee_bench_ts_global);
+
+		optee_bench_ts_global = NULL;
+
```
style: empty line; here and before each `break;` above.

jforissier commented:
The commit subject should probably be 'tee: optee: add benchmark infrastructure **not for maiinline**`, and the commit text should describe why we're adding this (just a few words and maybe Link: to the optee_os commit once it's merged?)

igoropaniuk commented:
@jforissier fixed

etienne-lms commented on drivers/tee/optee/bench.c:
```diff
@@ -0,0 +1,153 @@
+
+static inline u64 read_ccounter(void)
+{
+	u64 ccounter = 0;
```
useless init

igoropaniuk replied:
fixed

etienne-lms commented on drivers/tee/optee/bench.c:
```diff
@@ -0,0 +1,150 @@
+		return;
+
+	cur_cpu = get_cpu();
+	/* fill timestamp data */
```
minor style: useless comment.

igoropaniuk replied:
fixed

etienne-lms commented on drivers/tee/optee/bench.c:
```diff
@@ -0,0 +1,150 @@
+	int cur_cpu = 0;
+	int ret;
+
+	if (!optee_bench_ts_global)
```
Mutex required here in case `optee_bench_ts_global` gets deregistered ?

igoropaniuk replied:
@etienne-lms 
right, thanks. 
BTW I think it's better to put here kind of rwlock (to avoid blocking of other bm_timestamp invocations)

igoropaniuk replied:
@etienne-lms, please check last commit with fix

igoropaniuk commented:
I think I've addressed all comments so far and now think this is ready for a final review and if there are no more comments we should consider merging this now

jockebech commented on drivers/tee/optee/bench.c:
```diff
@@ -0,0 +1,157 @@
+ * Specific defines for ARM performance timers
+ */
+/* aarch32 */
+#define OPTEE_BENCH_DEF_OPTS			(1 | 16)
```
This is OK, but in many places in kernel they are using `BIT(nr)` macro (`./include/linux/bitops.h`).

jockebech commented on drivers/tee/optee/bench.c:
```diff
@@ -0,0 +1,157 @@
+
+/* aarch64 */
+#define OPTEE_BENCH_ARMV8_PMCR_MASK	0x3f
+#define OPTEE_BENCH_ARMV8_PMCR_E	(1 << 0) /* Enable all counters */
```
Actually here and below it probably makes sense to use the `BIT` macro.

jockebech commented on drivers/tee/optee/bench.c:
```diff
@@ -0,0 +1,157 @@
+#else
+	/* Enable EL0 access to PMU counters */
+	asm volatile("mcr p15, 0, %0, c9, c14, 0" :: "r"(1));
+	/* Enable all PMU counters */
```
Is this comment correct? It's about setting the divider?

jockebech commented on drivers/tee/optee/call.c:
```diff
@@ -144,10 +145,14 @@ u32 optee_do_call_with_arg(struct tee_context *ctx, phys_addr_t parg)
 	while (true) {
 		struct arm_smccc_res res;
 
+		optee_bm_timestamp();
```
I think we shall in the future discuss whether to have this here by default or not. There are pros and cons to actually have the timestamp calls pre populated in the code. It's nice to have "some" default data when enabling the framework. But if I compare with ftrace for example, there is nothing printed/save just by enabling the feature. What will be measured depends on what the user want to see. But again, let's take that in future patches.

jockebech commented:
With or without my BIT suggestions:
`Reviewed-by: Joakim Bech <joakim.bech@linaro.org>`

etienne-lms commented on drivers/tee/optee/Kconfig:
```diff
@@ -5,3 +5,10 @@ config OPTEE
 	  driver.
+
+config OPTEE_BENCHMARK
+	bool "OP-TEE Benchmark"
```
I suggest to add `(EXPERIMENTAL)` to the description as it is not fully a mature today.

igoropaniuk replied:
fixed

etienne-lms commented on drivers/tee/optee/optee_msg.h:
```diff
@@ -414,5 +414,5 @@ struct optee_msg_arg {
  */
 #define OPTEE_MSG_RPC_CMD_SHM_FREE	7
-
+#define OPTEE_MSG_RPC_CMD_BENCH_REG	20
```
Add a short description above, with same formating as for the other command IDs.

igoropaniuk replied:
fixed

etienne-lms commented on drivers/tee/optee/rpc.c:
```diff
@@ -307,6 +309,48 @@ static void handle_rpc_func_cmd_shm_free(struct tee_context *ctx,
+static void handle_rpc_func_cmd_bm_reg(struct optee_msg_arg *arg)
+{
+	u64 size;
+	u64 type;
```
minor: i don't think these variables are really useful.

igoropaniuk replied:
@etienne-lms 
size is used here (line 330):
```
memremap(arg->params[0].u.value.b, size, MEMREMAP_WB);
```
so I thought that it's better to have `size` than `arg->params[0].u.value.c;` in func params (for the sake of better readability)

the same for `type`

etienne-lms replied:
ok. maybe you could do the same for `paddr = arg->params[0].u.value.b`.

igoropaniuk replied:
@etienne-lms right, thanks!
fixed

etienne-lms commented on drivers/tee/optee/optee_bench.h:
```diff
@@ -0,0 +1,65 @@
+#include <linux/rwsem.h>
+#include <linux/tee_drv.h>
+
+#define OPTEE_BENCH_DIVIDER			64
```
Add same comment as in optee_os:
```
/*
 * Cycle count divider is enabled (in PMCR),
 * CCNT value is incremented every 64th clock cycle
 */

igoropaniuk replied:
fixed

etienne-lms commented on drivers/tee/optee/optee_msg.h:
```diff
@@ -414,5 +414,14 @@ struct optee_msg_arg {
  *					above
  */
 #define OPTEE_MSG_RPC_CMD_SHM_FREE	7
+/*
```
minor: add empty line before comment

etienne-lms commented on drivers/tee/optee/optee_msg.h:
```diff
@@ -414,5 +414,14 @@ struct optee_msg_arg {
+ * [in] param[0].u.value.a	Subcommand (register buffer, unregister buffer)
+ * [in] param[0].u.value.b	Physical address of timestamp buffer
+ * [in] param[0].u.value.c	Size of buffer
+ *
```
minor: remove empty comment line

igoropaniuk replied:
fixed

etienne-lms commented on drivers/tee/optee/optee_msg.h:
```diff
@@ -414,5 +414,14 @@ struct optee_msg_arg {
+ * [in] param[0].u.value.c	Size of buffer
+ *
+ */
+
```
minor: remove empty line

etienne-lms reviewed (commented):
empty line minor issues...

igoropaniuk commented on drivers/tee/optee/rpc.c:
```diff
@@ -307,6 +309,48 @@ static void handle_rpc_func_cmd_shm_free(struct tee_context *ctx,
+static void handle_rpc_func_cmd_bm_reg(struct optee_msg_arg *arg)
+{
+	u64 size;
+	u64 type;
```
fixed
